pkg/trace/info/stats.go

```diff
@@ -155,15 +169,127 @@ func (ts *TagStats) publish() {
 	metrics.Count("datadog.trace_agent.receiver.events_extracted", eventsExtracted, tags, 1)
 	metrics.Count("datadog.trace_agent.receiver.events_sampled", eventsSampled, tags, 1)
 	metrics.Count("datadog.trace_agent.receiver.payload_accepted", requestsMade, tags, 1)
+
+	for reason, count := range ts.TracesDropped.toMap() {
+		metrics.Count("datadog.trace_agent.normalizer.traces_dropped", count, append(tags, "reason:"+reason), 1)
+	}
+
+	for reason, count := range ts.TracesMalformed.toMap() {
+		metrics.Count("datadog.trace_agent.normalizer.traces_malformed", count, append(tags, "reason:"+reason), 1)
+	}
+}
+
+// statsStructToMap marshals the provided struct into a map[string]int64 by first marshaling it into json
+// then marshing from json into the map
+func statsStructToMap(v interface{}, name string) (result map[string]int64) {
```

Comment:
@gbbr, what do you think of this? It's kind of hacky... but it is a nice way to have the source of truth for the normalization `reason` names be in one place, the json-annotations, and then we use that everywhere else without having to duplicate: reason tags for metrics, agent info page, periodic log message.

Reply:
I agree. This _is_ quite hacky. The thing that you are trying to save here (one place for tags) is not worth the cost of the complexity and time it takes a reader to try and figure out this code. Furthermore, it also adds JSON struct fields to the type, creating even more confusion, leading people to believe that this is marshaled (which it is, but for the wrong reasons). 

Honestly, it also crossed my mind that it would be nice to have struct fields which represent the tags that these stats belong to. And this is doable, with custom struct fields, but I'm not sure if it's worth the effort. And even if we were to do it, then we should be doing it for all fields and simply include the metric and its tags.

If you like, I can suggest two ways to do this which are more "Go-like". 

1. Keep one publish method on each of these structs, which publishes the metrics just like the already existing one. For example:
```go
// publish publishes and resets the accumulated metrics.
func (td *TracesDropped) publish(tags []string) {
	metrics.Count("datadog.trace_agent.normalizer.traces_dropped", atomic.SwapInt64(&td.DecodingError), append(tags, "reason:decoding_error"), 1)
	metrics.Count("datadog.trace_agent.normalizer.traces_dropped", atomic.SwapInt64(&td.EmptyTrace), append(tags, "reason:empty_trace"), 1)
	metrics.Count("datadog.trace_agent.normalizer.traces_dropped", atomic.SwapInt64(&td.TraceIDZero), append(tags, "reason:trace_id_zero"), 1)
	metrics.Count("datadog.trace_agent.normalizer.traces_dropped", atomic.SwapInt64(&td.SpanIDZero), append(tags, "reason:span_id_zero"), 1)
	metrics.Count("datadog.trace_agent.normalizer.traces_dropped", atomic.SwapInt64(&td.ForeignSpan), append(tags, "reason:foreign_span"), 1)
}
```
This makes it clear to everyone because it behaves just like `(*TagStats).publish`. It would then be called there and receive the additional tags.

2. Use the suggestion given [here](https://github.com/DataDog/datadog-agent/pull/3738#discussion_r296569285). Personally, I prefer this one because it is very linear and easy to follow.

Lastly, I think a good practice in Go would be to have all these tags as constants, for example:
```go
const (
	tagReasonDuplicateSpanID       = "reason:duplicate_span_id"
	tagReasonServiceEmpty          = "reason:service_empty"
	tagReasonServiceTruncate       = "reason:service_truncate"
	tagReasonServiceInvalid        = "reason:service_invalid"
	tagReasonSpanNameEmpty         = "reason:span_name_empty"
	tagReasonSpanNameTruncate      = "reason:span_name_truncate"
	tagReasonSpanNameInvalid       = "reason:span_name_invalid"
	tagReasonResourceEmpty         = "reason:resource_empty"
	tagReasonTypeTruncate          = "reason:type_truncate"
	tagReasonInvalidStartDate      = "reason:invalid_start_date"
	tagReasonInvalidDuration       = "reason:invalid_duration"
	tagReasonInvalidHTTPStatusCode = "reason:invalid_http_status_code"
	tagReasonDecodingError         = "reason:decoding_error"
	tagReasonEmptyTrace            = "reason:empty_trace"
	tagReasonTraceIDZero           = "reason:trace_id_zero"
	tagReasonSpanIDZero            = "reason:span_id_zero"
	tagReasonForeignSpan           = "reason:foreign_span"
)
```

There is nothing wrong with being verbose if it helps with clarity. I find any of the above much easier to understand.

Reply:
> Lastly, I think a good practice in Go would be to have all these tags as constants, for example:

That seems unnecessarily verbose. I think as long as the reason names are used only in one place then we don't need to do that? 

Following-up to the slack discussion we had I made several changes:

* replaced hacky JSON with `tagValues()` method on each struct, which returns a map of `reason_name -> count`. That map is then used for sending metrics and for generating the concise string representation for logs
* now we send counters for all stats even if count=0

Reply:
Sure. SGTM